### PR TITLE
Testkit and RPC tests refactor

### PR DESCRIPTION
**Pull request by torkelrogstad**
Refactor testkit into returning promises, avoid all blocking in RPC tests

### CODE REVIEW

**Comment by Christewart** [comment]
When running this locally i did receive one failure: 

```
[info] - should be able to prune the blockchain *** FAILED ***
[info]   Condition timed out after 50 attempts with 3 seconds waiting periods (BitcoindRpcTestUtil.scala:390)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.bitcoins.rpc.BitcoindRpcTestUtil.awaitSynced(BitcoindRpcTestUtil.scala:390)
[info]   at org.bitcoins.rpc.BitcoindRpcTestUtil.awaitSynced$(BitcoindRpcTestUtil.scala:375)
[info]   at org.bitcoins.rpc.BitcoindRpcTestUtil$.awaitSynced(BitcoindRpcTestUtil.scala:805)
[info]   at org.bitcoins.rpc.BitcoindRpcTestUtil.$anonfun$generateAndSync$3(BitcoindRpcTestUtil.scala:368)
[info]   at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
[info]   at scala.collection.Iterator.foreach(Iterator.scala:941)
[info]   at scala.collection.Iterator.foreach$(Iterator.scala:941)
[info]   at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
```

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/BitcoindInstanceTest.scala [hunk 1]
```diff
@@ -42,18 +41,13 @@ class BitcoindInstanceTest extends AsyncFlatSpec with BeforeAndAfterAll {
   it should "parse a bitcoin.conf file, start bitcoind, mine some blocks and quit" in {
     val instance = BitcoindInstance.fromDatadir(datadir.toFile)
     val client = new BitcoindRpcClient(instance)
-    BitcoindRpcTestUtil.startServers(Vector(client))
-    RpcUtil.awaitServer(client)
 
     for {
+      _ <- BitcoindRpcTestUtil.startServers(Vector(client))
```
Comment (Christewart) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/BitcoindInstanceTest.scala]: We should probably implement helper methods so we don't have to wrap these in `Vector()`, i.e. `BitcoindRpcTestUtil.startServer` and `BitcoindRpcTestUtil.stopServer`

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/RpcUtilTest.scala [hunk 2]
```diff
@@ -15,15 +16,13 @@ import scala.util.Success
 
 class RpcUtilTest extends AsyncFlatSpec with BeforeAndAfterAll {
```
Comment (Christewart) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/RpcUtilTest.scala:17]: can all of these test cases `extends BitcoinSUnitTest`?

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/common/BlockchainRpcTest.scala [hunk 3]
```diff
@@ -1,302 +1,357 @@
...
+      _ <- freshClient
+        .addNode(otherFreshClient.getDaemon.uri, AddNodeArgument.OneTry)
+      _ <- AsyncUtil.retryUntilSatisfiedF(() =>
+        BitcoindRpcTestUtil.hasSeenBlock(otherFreshClient, bestHash1))
+
+      _ <- otherFreshClient.preciousBlock(bestHash1)
+      newBestHash <- otherFreshClient.getBestBlockHash
+
+    } yield assert(newBestHash == bestHash1)
+    /*BitcoindRpcTestUtil.createNodePair().flatMap {
```
Comment (Christewart) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/common/BlockchainRpcTest.scala]: ping

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/common/Client.scala [hunk 4]
```diff
@@ -88,26 +103,45 @@ trait Client {
...
+          s"Could not start bitcoind instance! Message: ${exc.getMessage}")
+        logger.info(s"Log lines:")
+        val logLines = Source.fromFile(logFile).getLines()
+        logLines.foreach(logger.info)
+    }
 
     started
   }
 
   def isStarted: Boolean = {
```
Comment (Christewart) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/common/Client.scala]: probably should get rid of this method

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/AsyncUtil.scala [hunk 5]
```diff
@@ -101,45 +110,51 @@ abstract class AsyncUtil extends BitcoinSLogger {
...
     */
   def awaitCondition(
       condition: () => Boolean,
-      duration: FiniteDuration = 100.milliseconds,
-      maxTries: Int = 50,
-      overallTimeout: FiniteDuration = 1.hour)(
-      implicit system: ActorSystem): Unit = {
+      duration: FiniteDuration = DEFAULT_DURATION,
+      maxTries: Int = DEFAULT_MAX_TRIES)(
+      implicit system: ActorSystem): Future[Unit] = {
```
Comment (Christewart) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/AsyncUtil.scala:124]: :tada:

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/AsyncUtil.scala [hunk 6]
```diff
@@ -101,45 +110,51 @@ abstract class AsyncUtil extends BitcoinSLogger {
...
 
   def awaitConditionF(
       conditionF: () => Future[Boolean],
-      duration: FiniteDuration = 100.milliseconds,
-      maxTries: Int = 50,
-      overallTimeout: FiniteDuration = 1.hour)(
-      implicit system: ActorSystem): Unit = {
+      duration: FiniteDuration = DEFAULT_DURATION,
+      maxTries: Int = DEFAULT_MAX_TRIES)(
+      implicit system: ActorSystem): Future[Unit] = {
```
Comment (Christewart) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/AsyncUtil.scala:139]: :tada:

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/AsyncUtil.scala [hunk 7]
```diff
@@ -101,45 +110,51 @@ abstract class AsyncUtil extends BitcoinSLogger {
...
   }
 }
 
-object AsyncUtil extends AsyncUtil
+object AsyncUtil extends AsyncUtil {
+
+  /**
+    * The default duration between async attempt
+    */
+  private[bitcoins] val DEFAULT_DURATION: FiniteDuration = 100.milliseconds
```
Comment (Christewart) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/AsyncUtil.scala]: `DEFAULT_INTERVAL` would be a better name

File: testkit/src/main/scala/org/bitcoins/rpc/BitcoindRpcTestUtil.scala [hunk 8]
```diff
@@ -41,13 +41,22 @@ import org.bitcoins.util.{AsyncUtil, ListUtil}
...
 import scala.concurrent._
 import scala.concurrent.duration.{DurationInt, FiniteDuration}
 import scala.util._
 
 //noinspection AccessorLikeMethodIsEmptyParen
 trait BitcoindRpcTestUtil extends BitcoinSLogger {
   import scala.collection.JavaConverters._
+  import BitcoindRpcTestUtil.DEFAULT_LONG_DURATION
+
+  type RpcClientAccumVersioned[T <: BitcoindRpcClient] =
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/rpc/BitcoindRpcTestUtil.scala:54]: Any reason not to make this just a case class?

File: testkit/src/main/scala/org/bitcoins/rpc/RpcUtil.scala [hunk 9]
```diff
@@ -1,28 +1,63 @@
...
 
     AsyncUtil.transformRetryToTestFailure(retryF)(system.dispatcher)
   }
+
+  def awaitServerShutdown(
+      server: BitcoindRpcClient,
+      duration: FiniteDuration = 300.milliseconds,
+      maxTries: Int = 50)(implicit system: ActorSystem): Future[Unit] = {
+    import system.dispatcher
+    val f = () => server.isStartedF.map(isStarted => !isStarted)
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/rpc/RpcUtil.scala]: might be worthwhile to implemet a `isStoppedF` in `Client`

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/RpcUtilTest.scala (hunk 2 above)
Comment (torkelrogstad) [reply bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/RpcUtilTest.scala:17]: `BitcoinSUnitTest` extends `FlatSpec`, RPC tests use `AsyncFlatSpec`. We could make a `BitcoinSAsyncUnitTest`?

File: testkit/src/main/scala/org/bitcoins/rpc/BitcoindRpcTestUtil.scala (hunk 8 above)
Comment (torkelrogstad) [reply testkit/src/main/scala/org/bitcoins/rpc/BitcoindRpcTestUtil.scala:54]: A case class that wraps a vectorbuilder? This type is to make sure users can pass in a builder of the approriate type without getting an error.

**Pull request merged**